protocol/protocol.typeHierarchy.proposed.md

```diff
@@ -0,0 +1,77 @@
+
+#### Type Hierarchy
+
+Many language support inheritance and LSP provides retrieving the hierarchy information with the following requests.
+
+
+The `textDocument/subTypes` request is sent from the client to the server to resolve resolve the subtypes of a symbol at a given text document position. Returns the direct subtypes in no particular order.
+
+_Client Capabilities_:
+
+```ts
+  TextDocumentClientCapabilities {
+    /**
+     * Capabilities specific to the `textDocument/subTypes`
+     *  and `textDocument/superTypes
+     */
+    typeHierarchy?: {
+    /**
+     * Whether implementation supports dynamic registration.
+     */
+    dynamicRegistration?: boolean;
+  };
+```
+
+_Server Capabilities_:
+```ts
+ServerCapabilities{
+  /**
+   * The server provides type hierarchy information
+   */
+  typeHierarchyProvider?: boolean
+}
+```
+
+```ts
+/**
+ * Represents hierarchical information about programming constructs like variables, classes,
+ * interfaces etc.
+ */
+export interface SymbolNode extends SymbolInformation {
+  /**
+   * Immediate descendants of this node
+   */
+  descendants? : SymbolNode[]
```

Comment:
If it's the immediate descendants, it should be called `children`

Reply:
The doc is incorrect needs to be updated. 

I am not sure about children either. This can actually hold hierarchies with depth larger than one. Probably needs a better name.

Reply:
Even if it's depth larger than one, wouldn't the elements of this particular array still be the immediate children?

Reply:
We use this same data structure for both `subTypes` and `superTypes` calls on the superTypes call they are actually the ancestors/parents. We probably need a new name to reflect that.

Reply:
It may sometimes be useful to expand multiple levels in one request, This is how [ccls](https://github.com/MaskRay/ccls/blob/master/src/messages/ccls_call_hierarchy.cc#L44) and cquery do.